File: src/main/java/com/pixelreel/service/provider/RawgGameSearchProvider.java
package com.pixelreel.service.provider;

import com.pixelreel.config.RawgProperties;
import com.pixelreel.dto.external.ExternalGameSearchResult;
import com.pixelreel.dto.external.GameRecordSuggestion;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.dto.rawg.RawgSearchResponse;
import com.pixelreel.entity.Game;
import com.pixelreel.enums.RecordStatus;
import com.pixelreel.service.GameService;
import com.pixelreel.service.search.GameSearchProvider;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class RawgGameSearchProvider implements GameSearchProvider {
  private static final int PAGE_SIZE = 20;

  private final GameService gameService;

  @Qualifier("rawgRestClient")
  private final RestClient rawgRestClient;

  private final RawgProperties rawgProperties;

  @Override
  public String id() {
    return "rawg";
  }

  @Override
  public ProviderSearchResult<ExternalGameSearchResult> search(String query, int page) {
    ProviderSearchResult<ExternalGameSearchResult> result = new ProviderSearchResult<>();
    result.setProvider(id());

    if (!rawgProperties.enabled()) {
      result.setEnabled(false);
      result.setMessage("RAWG 未启用");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(rawgProperties.apiKey())) {
      result.setEnabled(false);
      result.setMessage("缺少 RAWG API Key");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(query)) {
      throw new IllegalArgumentException("query must not be blank");
    }

    int normalizedPage = Math.max(page, 1);
    RawgSearchResponse response = rawgRestClient
        .get()
        .uri(uriBuilder -> uriBuilder
            .path("/games")
            .queryParam("search", query)
            .queryParam("page", normalizedPage)
            .queryParam("page_size", PAGE_SIZE)
            .queryParam("key", rawgProperties.apiKey())
            .build())
        .retrieve()
        .body(RawgSearchResponse.class);

    List<RawgSearchResponse.RawgGame> items = response == null || response.getResults() == null
        ? Collections.emptyList()
        : response.getResults();
    Map<Long, Game> existingByRawgId = findExistingByRawgId(items);

    List<ExternalGameSearchResult> results = new ArrayList<>();
    for (RawgSearchResponse.RawgGame item : items) {
      if (item == null || !StringUtils.hasText(item.getName())) {
        continue;
      }

      ExternalGameSearchResult mapped = new ExternalGameSearchResult();
      mapped.setProvider(id());
      mapped.setRawgId(item.getId());
      mapped.setTitle(item.getName());
      mapped.setPosterUrl(normalizeUrl(item.getBackgroundImage()));
      mapped.setReleaseDate(item.getReleased());

      Game existing = item.getId() == null ? null : existingByRawgId.get(item.getId());
      mapped.setAlreadyAdded(existing != null);
      mapped.setExistingRecordId(existing == null ? null : existing.getId());
      mapped.setSuggestedRecord(buildSuggestion(mapped));
      results.add(mapped);
    }

    int total = response == null || response.getCount() == null ? results.size() : response.getCount();
    int totalPages = total == 0 ? 0 : (int) Math.ceil(total / (double) PAGE_SIZE);

    result.setEnabled(true);
    result.setPage(normalizedPage);
    result.setTotalPages(totalPages);
    result.setTotalResults(total);
    result.setResults(results);
    return result;
  }

  private Map<Long, Game> findExistingByRawgId(List<RawgSearchResponse.RawgGame> items) {
    List<Long> ids = items.stream()
        .map(RawgSearchResponse.RawgGame::getId)
        .filter(value -> value != null)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return gameService.lambdaQuery()
        .in(Game::getRawgId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Game::getRawgId, Function.identity(), (a, b) -> a));
  }

  private GameRecordSuggestion buildSuggestion(ExternalGameSearchResult mapped) {
    GameRecordSuggestion suggestion = new GameRecordSuggestion();
    suggestion.setRawgId(mapped.getRawgId());
    suggestion.setTitle(mapped.getTitle());
    suggestion.setPosterUrl(mapped.getPosterUrl());
    suggestion.setStatus(RecordStatus.WANT);
    suggestion.setRating(null);
    suggestion.setShortReview("");
    return suggestion;
  }

  private String normalizeUrl(String url) {
    if (!StringUtils.hasText(url)) {
      return null;
    }
    String trimmed = url.trim();
    if (trimmed.startsWith("//")) {
      return "https:" + trimmed;
    }
    return trimmed;
  }
}

File: express-backend/src/services/provider/RawgGameSearchProvider.ts
import axios from 'axios';
import { config } from '../../config';
import { getDb } from '../../config/db';
import { GameSearchProvider } from '../provider/GameSearchProvider';
import {
  ExternalGameSearchResult,
  ProviderSearchResult,
  GameRecordSuggestion,
} from '../../dto/external-search';
import { RecordStatus } from '../../enums/RecordStatus';

function containsChinese(str: string): boolean {
  return /[一-鿿]/.test(str);
}

// MyMemory 免费翻译 API（无需 key）
async function translateToEnglish(text: string): Promise<string | null> {
  try {
    const res = await axios.get('https://api.mymemory.translated.net/get', {
      params: { q: text, langpair: 'zh|en' },
      timeout: 5000,
    });
    const translated = res.data?.responseData?.translatedText;
    if (translated && translated.toLowerCase() !== text.toLowerCase()) {
      return translated;
    }
    return null;
  } catch {
    return null;
  }
}

// RAWG 游戏搜索 Provider，与 Java 端 RawgGameSearchProvider 完全对齐
export class RawgGameSearchProvider implements GameSearchProvider {
  private static readonly PAGE_SIZE = 20;

  id(): string {
    return 'rawg';
  }

  async search(query: string, page: number): Promise<ProviderSearchResult<ExternalGameSearchResult>> {
    const result: ProviderSearchResult<ExternalGameSearchResult> = {
      provider: this.id(),
      enabled: true,
      message: '',
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };

    if (!config.rawg.apiKey) {
      result.enabled = false;
      result.message = '缺少 RAWG API Key';
      return result;
    }

    if (!query) throw new Error('query must not be blank');
    const normalizedPage = Math.max(page, 1);

    // RAWG 不支持中文搜索，检测到中文时翻译为英文
    let searchQuery = query;
    if (containsChinese(query)) {
      const translated = await translateToEnglish(query);
      if (translated) {
        searchQuery = translated;
      }
    }

    const response = await axios.get(`${config.rawg.baseUrl}/games`, {
      params: {
        search: searchQuery,
        page: normalizedPage,
        page_size: RawgGameSearchProvider.PAGE_SIZE,
        key: config.rawg.apiKey,
      },
    });

    const items = response.data?.results ?? [];
    const rawgIds = items.map((i: any) => i.id).filter(Boolean);
    const existingMap = rawgIds.length > 0
      ? await this.findExistingByRawgId(rawgIds)
      : new Map<any, any>();

    const results: ExternalGameSearchResult[] = [];
    for (const item of items) {
      if (!item || !item.name) continue;
      const existing = item.id ? existingMap.get(item.id) ?? null : null;
      const posterUrl = item.background_image
        ? (item.background_image.startsWith('//') ? 'https:' + item.background_image : item.background_image)
        : null;
      const mapped: ExternalGameSearchResult = {
        provider: this.id(),
        rawgId: item.id ?? null,
        steamAppId: null,
        xboxId: null,
        psnId: null,
        title: item.name,
        posterUrl,
        releaseDate: item.released ?? null,
        overview: null,
        alreadyAdded: existing !== null,
        existingRecordId: existing?.id != null ? Number(existing.id) : null,
        suggestedRecord: null,
      };
      mapped.suggestedRecord = this.buildSuggestion(mapped);
      results.push(mapped);
    }

    const total = response.data?.count ?? results.length;
    const totalPages = total === 0 ? 0 : Math.ceil(total / RawgGameSearchProvider.PAGE_SIZE);

    result.page = normalizedPage;
    result.totalPages = totalPages;
    result.totalResults = total;
    result.results = results;
    return result;
  }

  private async findExistingByRawgId(ids: number[]): Promise<Map<any, any>> {
    const games = await getDb().game.findMany({ where: { rawgId: { in: ids } } });
    return new Map(games.map((g) => [g.rawgId!, g]));
  }

  private buildSuggestion(mapped: ExternalGameSearchResult): GameRecordSuggestion {
    return {
      rawgId: mapped.rawgId,
      steamAppId: mapped.steamAppId,
      xboxId: mapped.xboxId,
      psnId: mapped.psnId,
      title: mapped.title,
      posterUrl: mapped.posterUrl,
      status: RecordStatus.WANT,
      rating: null,
      shortReview: '',
    };
  }
}